### DISCUSSION TITLE
How can I hide my js code from the user?

### DISCUSSION THREAD

**Discussion by chikashimiyama** [nick-thompson/react-juce#268] [Q&A]
In the react-juce example, C++ code loads the js code from the file system

> auto bundle = sourceDir.getChildFile("jsui/build/js/main.js");
> appRoot_.evaluate(bundle);

This is not ideal for a commercial product because the user can read, delete or modify the code intentionally or accidentally.
Would it be possible to embed the js file completely in the plugin? for example, using juce's binary data?

**Comment by nick-thompson** [comment]
Good question, and yes of course! Just as you suggested, once you've built your `main.js` you can add it to your JUCE project as binary data, then read and evaluate your code from memory at runtime. This is exactly what [Remnant](https://www.creativeintent.co/product/remnant) does.

To do that, you'll want to manage your own `EcmascriptEngine` and interface with it directly:

```cpp
std::shared_ptr<reactjuce::EcmascriptEngine> engine = std::make_shared<reactjuce::EcmascriptEngine>();
reactjuce::ReactApplicationRoot appRoot (engine);

addAndMakeVisible(appRoot);
engine->evaluateInline(juce::String(juce::CharPointer_UTF8(juce::BinaryData::main_js)));
```

**Reply by JoshMarler** [reply]
Yeah. FWIW we're using the Hermes engine and as part of our build process. For release we compile the bundled js file to Hermes bytecode and load that directly, similar to the above approach. It gives super fast load times etc.

I've been meaning to document an approach for doing this for a while. I'll try to get something submitted to the docs this weekend.
